numba/typeddict.py

```diff
@@ -0,0 +1,200 @@
+"""
+Python wrapper that connects CPython interpreter to the numba dictobject.
+"""
+from collections import MutableMapping
+
+from numba.types import DictType, TypeRef
+from numba import njit, dictobject, types, cgutils
+from numba.extending import (
+    overload_method,
+    box,
+    unbox,
+    NativeValue
+)
+
+
+@njit
+def _make_dict(keyty, valty):
+    return dictobject._as_meminfo(dictobject.new_dict(keyty, valty))
+
+
+@njit
+def _length(d):
+    return len(d)
+
+
+@njit
+def _setitem(d, key, value):
+    d[key] = value
+
+
+@njit
+def _getitem(d, key):
+    return d[key]
+
+
+@njit
+def _delitem(d, key):
+    del d[key]
+
+
+@njit
+def _contains(d, key):
+    return key in d
+
+
+@njit
+def _get(d, key, default):
+    return d.get(key, default)
+
+
+@njit
+def _setdefault(d, key, default):
+    return d.setdefault(key, default)
+
+
+@njit
+def _iter(d):
+    return list(d.keys())
+
+
+def _from_meminfo_ptr(ptr, dicttype):
+    d = TypedDict(meminfo=ptr, dcttype=dicttype)
+    return d
+
+
+class TypedDict(MutableMapping):
+    """A typed-dictionary usable in Numba.
+
+    Implements the MutableMapping interface.
+    """
+    @classmethod
+    def empty(cls, key_type, value_type):
+        """Create a new empty TypedDict with *key_type* and *value_type*
+        as the types for the keys and values of the dictionary.
+        """
+        return cls(dcttype=DictType(key_type, value_type))
+
+    def __init__(self, **kwargs):
+        """
+        Parameters
+        ----------
+        dcttype : numba.types.DictType; keyword-only
+            The dictionary type
+        meminfo : MemInfo; keyword-only
+            Used internally to pass the MemInfo object when boxing.
+        """
+        self._dict_type, self._opaque = self._parse_arg(**kwargs)
+
+    def _parse_arg(self, dcttype, meminfo=None):
+        if not isinstance(dcttype, DictType):
+            raise TypeError('*dcttype* must be a DictType')
+
+        if meminfo is not None:
+            opaque = meminfo
+        else:
+            opaque = _make_dict(dcttype.key_type, dcttype.value_type)
+        return dcttype, opaque
+
+    @property
+    def _numba_type_(self):
+        return self._dict_type
+
+    def __getitem__(self, key):
+        return _getitem(self, key)
+
+    def __setitem__(self, key, value):
+        return _setitem(self, key, value)
+
+    def __delitem__(self, key):
+        _delitem(self, key)
+
+    def __iter__(self):
+        return iter(_iter(self))
+
+    def __len__(self):
+        return _length(self)
+
+    def __contains__(self, key):
+        return _contains(self, key)
+
+    def __repr__(self):
```

Comment:
Perhaps this should be the `__str__` and the `__repr__` also contain some type information like `str(self._dict_type)` as a prefix ?

Reply:
good pt

Reply:
changed